addons/purchase_stock/models/purchase.py

```diff
@@ -103,6 +103,14 @@ def button_cancel(self):
                         if all(state in ('done', 'cancel') for state in siblings_states):
                             move_dest_ids.write({'procure_method': 'make_to_stock'})
                             move_dest_ids._recompute_state()
+                        if order_line.propagate:
```

Comment:
if propagate cancel:
   cancel
elif all(state in ('done', 'cancel') for state in siblings_states):
   recompute and write

avoid to compute the state and procure_method if you directly cancel it after